msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenRequestExecutor.java

```diff
@@ -86,25 +86,76 @@ private void addQueryParameters(OAuthHttpRequest oauthHttpRequest) {
         String clientID = msalRequest.application().clientId();
         queryParameters.put("client_id", clientID);
 
-        // If the client application has a client assertion to apply to the request, check if a new client assertion
-        //  was supplied as a request parameter. If so, use the request's assertion instead of the application's
+        // Add client authentication parameters if this is a confidential client
         if (msalRequest.application() instanceof ConfidentialClientApplication) {
-            if (msalRequest instanceof ClientCredentialRequest && ((ClientCredentialRequest) msalRequest).parameters.clientCredential() != null) {
-                IClientCredential credential = ((ClientCredentialRequest) msalRequest).parameters.clientCredential();
-                addJWTBearerAssertionParams(queryParameters, ((ConfidentialClientApplication) msalRequest.application()).getAssertionString(credential));
-            } else {
-                if (((ConfidentialClientApplication) msalRequest.application()).assertion != null) {
-                    addJWTBearerAssertionParams(queryParameters, ((ConfidentialClientApplication) msalRequest.application()).assertion);
-                } else if (((ConfidentialClientApplication) msalRequest.application()).secret != null) {
-                    // Client secrets have a different parameter than bearer assertions
-                    queryParameters.put("client_secret", ((ConfidentialClientApplication) msalRequest.application()).secret);
-                }
-            }
+            ConfidentialClientApplication application = (ConfidentialClientApplication) msalRequest.application();
+
+            // Determine which credential to use - either from the request or from the application
+            IClientCredential credential = getCredentialToUse(application);
+
+            // Add appropriate authentication parameters based on the credential type
+            addCredentialToRequest(queryParameters, credential, application);
         }
 
         oauthHttpRequest.setQuery(StringHelper.serializeQueryParameters(queryParameters));
     }
 
+    /**
+     * Determines which credential to use for authentication:
+     * - If the request is a ClientCredentialRequest with a specified credential, use that
+     * - Otherwise use the application's credential
+     *
+     * @param application The confidential client application
+     * @return The credential to use, may be null if no credential is available
+     */
+    private IClientCredential getCredentialToUse(ConfidentialClientApplication application) {
+        if (msalRequest instanceof ClientCredentialRequest &&
```

Comment:
Is this flow covered by tests? If not can we add tests to cover this?

Reply:
There are some end-to-end tests in ClientCredentialsIT that would go through this code path but it wasn't covered that well.

In the latest commit I added a new `testCredentialPrecedenceAndMixing()` unit test to `ClientCredentialTest` to more explicitly show the app-level credential getting overridden at the request level.